lib/popover.js

```diff
@@ -0,0 +1,311 @@
+};
+
+const PopoverInner = styled.div`
+  position: relative;
+  overflow: hidden;
+  background-color: var(--colors-background);
+  font-size: var(--fontSizes-small);
+  border-radius: var(--rounded);
+
+  border: 1px solid var(--colors-border);
+  box-shadow: var(--popShadow);
+  z-index: 1;
+`;
+
+const PopoverContent = ({ align, children, ...props }) => {
+  const { ref, offset } = usePositionAdjustment();
+  return (
+    <PopoverWrap data-module="Popover" align={align} {...props}>
+      <PopoverInner ref={ref} style={offset}>
+        {children}
+      </PopoverInner>
+    </PopoverWrap>
+  );
+};
+
+const PopoverContainer = styled.details`
+  position: relative;
+  display: inline-block;
+`;
+const PopoverToggle = styled.summary`
+  list-style-image: none;
+  &::-webkit-details-marker {
+    display: none;
+    background: none;
+    color: transparent;
+    width: 0;
+  }
+`;
+
+const useClickOutside = (open, onClickOutside) => {
+  const ref = React.useRef();
+  const onClickOutsideProxy = useCallbackProxy(onClickOutside);
+  React.useEffect(() => {
+    if (!open) return;
+    const handler = (e) => {
+      if (document.body.contains(e.target) && !ref.current.contains(e.target)) {
+        onClickOutsideProxy(e);
+      }
+    };
+    window.addEventListener('click', handler);
+    return () => {
+      window.removeEventListener('click', handler);
+    };
+  }, [open, onClickOutsideProxy]);
+  return ref;
+};
+
+const useStack = (defaultState) => {
+  const [stack, setStack] = React.useState([defaultState]);
+  const top = stack[stack.length - 1];
+  const push = (value) => setStack((prev) => prev.concat([value]));
+  const pop = () => setStack((prev) => prev.slice(0, -1));
+  const clear = () => setStack([defaultState]);
+  return { top, push, pop, clear };
+};
+
+export const Popover = ({ align, label, views = {}, initialView, children, contentProps, ...props }) => {
+  const [open, setOpen] = React.useState(false);
+  const focusedOnMount = React.useRef();
+  const { top: activeView, push: setActiveView, pop: onBack, clear } = useStack(initialView);
+  const activeViewFunc = views[activeView] || children;
+  const onOpen = () => {
+    setOpen(true);
+    clear();
+  };
+  const onClose = () => {
+    setOpen(false);
+    clear();
+  };
+  React.useEffect(() => {
+    if (open && focusedOnMount.current) focusedOnMount.current.focus();
+  }, [open, activeView]);
+
+  const rootRef = useClickOutside(open, onClose);
+
+  const onToggle = (e) => {
+    // NOTE: this runs _after_ it toggles; we're updating the React state so that it matches the internal state
+    if (e.target.open) {
+      onOpen();
+    } else {
+      onClose();
+    }
+  };
+
+  useEscape(onClose);
+
+  // see https://www.w3.org/TR/wai-aria-practices-1.1/#menubutton
+  return (
+    <PopoverContainer data-module="PopoverMenu" ref={rootRef} onToggle={onToggle} open={open} role="menu" {...props}>
+      <PopoverToggle aria-haspopup="menu" aria-expanded={open}>
```

Comment:
Can't find the implementation of this but - why is it a `<details>` element? It should be a button.

Reply:
hmm is that strictly true? a details element is a fully accessible way of revealing more content on the page, which seems to be the definition of a popover

Reply:
This is a technique I learned from [details on details](https://docs.google.com/presentation/d/1hvnPpsJo44BTPfJx28CV95vqk_dt6na1awUbk0kmZYM/edit#slide=id.g3e31444916_1_1) [[notes](https://github.com/muan/details-on-details)]

Reply:
Around 25% of screenreader users use Internet explorer as their primary browser, and `<details>` isn't supported at all in IE. Some sources suggest that means the details element contents will be constantly visible as expanded in IE.

I tried testing what it looks like in IE, but the app is using generators which aren't supported either so the whole page doesn't load at all.

Are we dropping support for IE?

Reply:
As far as I know, neither the editor nor community work properly in IE.

That said, if a consumer of this library wants to support IE, it can transpile this library via Babel and use this [polyfill](https://github.com/javan/details-element-polyfill) for the `<details>` element. 

Browser support strategies are [discussed a bit](https://github.com/FogCreek/shared-components/blob/master/README.md#browser-support) in the readme.

Reply:
What are we gaining by requiring a polyfill for the details element vs just using a button for this? 

The WAI-ARIA practices recommend using an anchor or a button element for a menu button and don't mention using `<details>` at all. I trust their recommendations since they've generally done more cross-browser and cross-AT testing.